File: solutions/cpp/binary-search/binary_search.cpp
#include "binary_search.h"

namespace binary_search
{
    using namespace std;

    size_t find(const vector<int> &list, const int key)
    {
        int low = 0, high = list.size() - 1;

        while (low <= high)
        {
            size_t mid = low + (high - low) / 2;
            int value = list[mid];

            if (value > key)
                high = mid - 1;
            else if (value < key)
                low = mid + 1;
            else
                return mid;
        }
        throw domain_error("not found");
    }
} // namespace binary_search

File: solutions/fortran/binary-search/binary_search.f90
module binary_search
   implicit none
contains

   function find(array, val) result(idx)
      integer, dimension(:), intent(in) :: array
      integer, intent(in) :: val
      integer :: idx, low, mid, high

      idx = -1
      low = 1
      high = size(array)

      do while (low <= high)
         mid = low + (high - low) / 2
         if ( array(mid) < val) then
            low = mid + 1
         else if ( array(mid) > val) then
            high = mid - 1
         else
            idx = mid
            exit
         end if
      end do

   end function

end module
